spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisQueueMessageDrivenEndpoint.java

```diff
@@ -272,7 +272,7 @@ private void restart() {
 	protected void doStop() {
 		try {
 			this.active = false;
-			this.lifecycleCondition.await(Math.min(this.stopTimeout, this.receiveTimeout), TimeUnit.MICROSECONDS);
+			this.lifecycleCondition.await(Math.min(this.stopTimeout, this.receiveTimeout), TimeUnit.MILLISECONDS);
```

Comment:
I think you might have misunderstood me; my point is that we really shouldn't block in `stop()` at all; we should be overriding `doStop(Runnable callback)` return, and invoke the callback after the listener has stopped.

Also, I noticed that the gateway is not invoking `super.doStop()` so the `replyMessageCorrelator` is not stopped. Same problem with `doStart()`.